ext/semian/semian_shared_memory_object.c

```diff
@@ -0,0 +1,629 @@
+#include "semian_shared_memory_object.h"
+
+const int kSHMSemaphoreCount = 1; // # semaphores to be acquired
+const int kSHMTicketMax = 1;
+const int kSHMInitializeWaitTimeout = 5; /* seconds */
+const int kSHMIndexTicketLock = 0;
+const int kSHMInternalTimeout = 5; /* seconds */
+
+static struct sembuf decrement; // = { kSHMIndexTicketLock, -1, SEM_UNDO};
+static struct sembuf increment; // = { kSHMIndexTicketLock, 1, SEM_UNDO};
+
+/*
+ * Functions that handle type and memory
+*/
+static void semian_shm_object_mark(void *ptr);
+static void semian_shm_object_free(void *ptr);
+static size_t semian_shm_object_memsize(const void *ptr);
+
+static void *semian_shm_object_lock_without_gvl(void *v_ptr);
+static void *semian_shm_object_unlock_without_gvl(void *v_ptr);
+
+const rb_data_type_t
+semian_shm_object_type = {
+  "semian_shm_object",
+  {
+    semian_shm_object_mark,
+    semian_shm_object_free,
+    semian_shm_object_memsize
+  },
+  NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+static void
+semian_shm_object_mark(void *ptr)
+{
+  /* noop */
+}
+static void
+semian_shm_object_free(void *ptr)
+{
+  semian_shm_object *data = (semian_shm_object *) ptr;
+  // Under normal circumstances, memory use should be in the order of bytes,
+  //   and shouldn't increase if the same key/id is used
+  //   so there is no need to call this unless certain all other semian processes are stopped
+  //   (also raises concurrency errors: "object allocation during garbage collection phase")
+
+  xfree(data);
+}
+static size_t
+semian_shm_object_memsize(const void *ptr)
+{
+  return
+
+  sizeof(semian_shm_object);
```

Comment:
This is technically valid, but please move this up to be on the same line as the `return` :-)